### PR TITLE
Do not log forge tokens

### PR DESCRIPTION
**Pull request by qwerty287**

### CODE REVIEW

File: server/forge/setup/setup.go [hunk 1]
```diff
@@ -64,7 +64,7 @@ func setupGitea(forge *model.Forge) (forge.Forge, error) {
 	if len(opts.URL) == 0 {
 		return nil, fmt.Errorf("WOODPECKER_GITEA_URL must be set")
 	}
-	log.Trace().Msgf("Forge (gitea) opts: %#v", opts)
+	log.Trace().Msg("Setting up gitea")
```
Comment (zc-devs) [inline server/forge/setup/setup.go:67]: 1. But it was a `trace` log, not even `debug`.
2. What if I want to log/debug forge parameters like URL, SkipVerify and OAuthHost?

Comment (zc-devs) [reply server/forge/setup/setup.go:67]: 1. Yeah, sure.
2. No :)
2.1. They are go through one transformation [at least](https://github.com/woodpecker-ci/woodpecker/blob/355ebcd508afe107cd5641c4eeb2e4e0c9474795/cmd/server/flags.go#L336). Which [might not work](https://github.com/woodpecker-ci/woodpecker/issues/4503).
2.2. The second transformation might be in some [setup logic](https://github.com/woodpecker-ci/woodpecker/blob/main/server/services/setup.go#L124-L159).
2.3. Also some settings can be changed via API, it appears
Screenshot 2024-12-10 1

**Comment by 6543** [comment]
trace log is fine to have secrets it's ment for debugging in development ...

**Comment by zc-devs** [comment]
JFYI, https://github.com/showa-93/go-mask?tab=readme-ov-file#field-name--map-key solves initial intention as well as my comment number 2.
Another solution could be copying the structure, deleting value of `Secret` filed and logging this redacted structure.

**Comment by pat-s** [comment]
Tokens should never be logged. At most you might want to care that they are not empty to ensure they didn't get lost in between. In this case you could print an abbreviated subset of the token.

**Comment by pat-s** [comment]
@qwerty287 This is currently the last issue in the milestone. Do you want to finish it or postpone? I don't think it is that critical that it has to go into 3.0 necessarily.

**Review by 6543 — APPROVED** [review]
as this specific part works fine I'm ok to limit it's verbosity to harden security ...

... still trace log is for devs or hard to catch bugs ...

File: server/forge/setup/setup.go [hunk 2]
```diff
@@ -64,7 +64,7 @@ func setupGitea(forge *model.Forge) (forge.Forge, error) {
 	if len(opts.URL) == 0 {
 		return nil, fmt.Errorf("WOODPECKER_GITEA_URL must be set")
 	}
-	log.Trace().Msgf("Forge (gitea) opts: %#v", opts)
+	log.Trace().Str("url", opts.URL).Str("oauth-host", opts.OAuthHost).Bool("skip-verify", opts.SkipVerify).Msg("Setting up gitea")
```
Comment (zc-devs) [inline server/forge/setup/setup.go:67]: Maybe `info` then? Like
```

{"level":"info","time":"2024-12-14T08:45:52Z","caller":"/woodpecker/src/github.com/woodpecker-ci/woodpecker/cmd/server/server.go:112","message":"starting Woodpecker server with version 'next-2e490f3c7b'"}
{"level":"info","time":"2024-12-14T08:45:52Z","caller":"/woodpecker/src/github.com/woodpecker-ci/woodpecker/cmd/server/server.go:117","message":"starting cron service ..."}
{"level":"info","time":"2024-12-14T08:45:52Z","caller":"/woodpecker/src/github.com/woodpecker-ci/woodpecker/cmd/server/server.go:128","message":"starting grpc server ..."}
{"level":"info","time":"2024-12-14T08:45:52Z","caller":"/woodpecker/src/github.com/woodpecker-ci/woodpecker/cmd/server/server.go:253","message":"starting http server ..."}
```

**Comment by xoxys** [comment]
Sorry, but why Info log level now? Why should we put that many details into info log?

**Comment by zc-devs** [comment]
IDK, if I may speak or I'm not who was asked. Hope, it won't be an offtopic.

> why Info log level now?

Because it would be aligned with [already existed logs](https://github.com/woodpecker-ci/woodpecker/pull/4551#discussion_r1884984561).

> Why should we put that many details into info log?

It's not necessary, though. I just suggested `info` first.
1. It could be two log lines: `info` with `starting...` and `debug` with details. 
2. Or else just one `debug` line.

**Comment by xoxys** [comment]
I would prefer debug for the additional logs.

**Comment by pat-s** [comment]
My vote goes for `debug` or `trace` log level incl. an abbreviated output of the token to ensure it is not empty (maybe 3 or 4 chars).

**Comment by xoxys** [comment]
What do you think about adding the masked token as well to at least signalize it is set? Something like:

```
logEvent := log.Debug().
    Str("url", opts.URL).
    Str("oauth-host", opts.OAuthHost).
    Bool("skip-verify", opts.SkipVerify)

if opts.Token != "" {
    logEvent = logEvent.Str("token", "********")
}

logEvent.Msg("Setting up gitea")
```

**Comment by xoxys** [comment]
Thats fine for me as well :) I have also seen strings like "_masked_" in logs of other tools but I dont have a strong opinion in how we signalize it.

**Review by xoxys — APPROVED** [review]

**Comment by xoxys** [comment]
> It's btw a similar problem for database connections, there we print the full data source as well which contains the password… I don't see an easy way to redact that though without completely omitting the data source

Good point 👍 lets look into it if we can find a solution, otherwise I agree we should drop it for now.

File: server/forge/setup/setup.go [hunk 3]
```diff
@@ -142,7 +142,7 @@ func setupBitbucketDatacenter(forge *model.Forge) (forge.Forge, error) {
 		Password:     gitPassword,
 		OAuthHost:    forge.OAuthHost,
 	}
-	log.Info().Str("url", opts.URL).Str("oauth-host", opts.OAuthHost).Msg("Setting up bitbucketdatacenter")
+	log.Debug().Str("url", opts.URL).Str("oauth-host", opts.OAuthHost).Msg("Setting up bitbucketdatacenter")
```
Comment (zc-devs) [inline server/forge/setup/setup.go]:
```suggestion
	log.Debug().Str("url", opts.URL).Str("oauth-host", opts.OAuthHost).Msg("Setting up Bitbucket Datacenter")
```

**Comment by xoxys** [comment]
Your suggestions can't be applied as you have done them based on an old commit. Instead of hard-coding the names, we should IMO reuse https://github.com/woodpecker-ci/woodpecker/blob/main/server/model/forge.go#L19-L27

File: server/forge/setup/setup.go [hunk 4]
```diff
@@ -44,7 +44,10 @@ func setupBitbucket(forge *model.Forge) (forge.Forge, error) {
 		Client: forge.Client,
 		Secret: forge.ClientSecret,
 	}
-	log.Trace().Msgf("Forge (bitbucket) opts: %#v", opts)
+	log.Debug().
+		Bool("client-set", opts.Client != "").
+		Bool("secret-set", opts.Secret != "").
+		Msg("Setting up bitbucket")
```
Comment (xoxys) [inline server/forge/setup/setup.go:50]:
```suggestion
		Msgf("Setting up forge %s", forge.Type)
```

File: server/forge/setup/setup.go [hunk 5]
```diff
@@ -64,7 +67,13 @@ func setupGitea(forge *model.Forge) (forge.Forge, error) {
...
 		return nil, fmt.Errorf("WOODPECKER_GITEA_URL must be set")
 	}
-	log.Trace().Msgf("Forge (gitea) opts: %#v", opts)
+	log.Debug().
+		Str("url", opts.URL).
+		Str("oauth-host", opts.OAuthHost).
+		Bool("skip-verify", opts.SkipVerify).
+		Bool("client-set", opts.Client != "").
+		Bool("secret-set", opts.Secret != "").
+		Msg("Setting up gitea")
```
Comment (xoxys) [inline server/forge/setup/setup.go:76]:
```suggestion
		Msgf("Setting up forge %s", forge.Type)
```

File: server/forge/setup/setup.go [hunk 6]
```diff
@@ -84,18 +93,32 @@ func setupForgejo(forge *model.Forge) (forge.Forge, error) {
...
 		return nil, fmt.Errorf("WOODPECKER_FORGEJO_URL must be set")
 	}
-	log.Trace().Msgf("Forge (forgejo) opts: %#v", opts)
+	log.Debug().
+		Str("url", opts.URL).
+		Str("oauth2-url", opts.OAuth2URL).
+		Bool("skip-verify", opts.SkipVerify).
+		Bool("client-set", opts.Client != "").
+		Bool("secret-set", opts.Secret != "").
+		Msg("Setting up forgejo")
```
Comment (xoxys) [inline server/forge/setup/setup.go:102]:
```suggestion
		Msgf("Setting up forge %s", forge.Type)
```

File: server/forge/setup/setup.go [hunk 7]
```diff
@@ -84,18 +93,32 @@ func setupForgejo(forge *model.Forge) (forge.Forge, error) {
...
 		OAuthHost:    forge.OAuthHost,
-	})
+	}
+	log.Debug().
+		Str("url", opts.URL).
+		Str("oauth-host", opts.OAuthHost).
+		Bool("skip-verify", opts.SkipVerify).
+		Bool("client-id-set", opts.ClientID != "").
+		Bool("client-secret-set", opts.ClientSecret != "").
+		Msg("Setting up gitlab")
```
Comment (xoxys) [inline server/forge/setup/setup.go:120]:
```suggestion
		Msgf("Setting up forge %s", forge.Type)
```

File: server/forge/setup/setup.go [hunk 8]
```diff
@@ -118,7 +141,15 @@ func setupGitHub(forge *model.Forge) (forge.Forge, error) {
...
-	log.Trace().Msgf("Forge (github) opts: %#v", opts)
+	log.Debug().
+		Str("url", opts.URL).
+		Str("oauth-host", opts.OAuthHost).
+		Bool("merge-ref", opts.MergeRef).
+		Bool("only-public", opts.OnlyPublic).
+		Bool("skip-verify", opts.SkipVerify).
+		Bool("client-set", opts.Client != "").
+		Bool("secret-set", opts.Secret != "").
+		Msg("Setting up github")
```
Comment (xoxys) [inline server/forge/setup/setup.go:152]:
```suggestion
		Msgf("Setting up forge %s", forge.Type)
```

File: server/forge/setup/setup.go [hunk 9]
```diff
@@ -140,16 +171,21 @@ func setupBitbucketDatacenter(forge *model.Forge) (forge.Forge, error) {
 		Password:     gitPassword,
 		OAuthHost:    forge.OAuthHost,
 	}
-	log.Trace().Msgf("Forge (bitbucketdatacenter) opts: %#v", opts)
+	log.Debug().
+		Str("url", opts.URL).
+		Str("oauth-host", opts.OAuthHost).
+		Bool("client-id-set", opts.ClientID != "").
+		Bool("client-secret-set", opts.ClientSecret != "").
+		Msg("Setting up bitbucketdatacenter")
```
Comment (xoxys) [inline server/forge/setup/setup.go:179]:
```suggestion
		Msgf("Setting up forge %s", forge.Type)
```

File: server/forge/setup/setup.go [hunk 10]
```diff
@@ -140,16 +171,21 @@ func setupBitbucketDatacenter(forge *model.Forge) (forge.Forge, error) {
...
 
 func setupAddon(forge *model.Forge) (forge.Forge, error) {
 	executable, ok := forge.AdditionalOptions["executable"].(string)
 	if !ok {
-		return nil, fmt.Errorf("missing git-username")
+		return nil, fmt.Errorf("missing addon executable")
 	}
 
-	log.Trace().Msgf("Forge (addon) executable: %#v", executable)
+	log.Debug().Msgf("Forge (addon) executable: %s", executable)
```
Comment (xoxys) [inline server/forge/setup/setup.go:189]:
```suggestion
	log.Debug().Str("executable", executable).Msgf("Setting up forge %s", forge.Type)
```

**Comment by xoxys** [comment]
Yes but it ensures we use the exact same spelling and name everywhere e.g. bitbucket-dc vs bitbucketdatacenter. However, If you prefer to keep the current approach fine for me as well.

**Comment by zc-devs** [comment]
I prefer it be human readable as I suggested.
Or else
```
.Str("type", forge.Type).Msg("Setting up forge")
```

**Comment by xoxys** [comment]
Adjusted the suggestions to use `.Str("type", forge.Type).Msg("Setting up forge")`

**Review by xoxys — APPROVED** [review]

**Review by xoxys — APPROVED** [review]

**Pull request merged**